// eslint-plugin-sonarjs/cjs/plugin-rules.js
"use strict";
/*
 * SonarQube JavaScript Plugin
 * Copyright (C) SonarSource Sàrl
 * mailto:info AT sonarsource DOT com
 *
 * You can redistribute and/or modify this program under the terms of
 * the Sonar Source-Available License Version 1, as published by SonarSource Sàrl.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.rules = void 0;
// DO NOT EDIT! This file was generated by generate-rule-indexes.ts
const index_js_1 = require("./S100/index.js");
const index_js_2 = require("./S101/index.js");
const index_js_3 = require("./S104/index.js");
const index_js_4 = require("./S105/index.js");
const index_js_5 = require("./S117/index.js");
const index_js_6 = require("./S124/index.js");
const index_js_7 = require("./S125/index.js");
const index_js_8 = require("./S126/index.js");
const index_js_9 = require("./S128/index.js");
const index_js_10 = require("./S134/index.js");
const index_js_11 = require("./S135/index.js");
const index_js_12 = require("./S138/index.js");
const index_js_13 = require("./S881/index.js");
const index_js_14 = require("./S888/index.js");
const index_js_15 = require("./S930/index.js");
const index_js_16 = require("./S1066/index.js");
const index_js_17 = require("./S1067/index.js");
const index_js_18 = require("./S1110/index.js");
const index_js_19 = require("./S1119/index.js");
const index_js_20 = require("./S1121/index.js");
const index_js_21 = require("./S1125/index.js");
const index_js_22 = require("./S1126/index.js");
const index_js_23 = require("./S1128/index.js");
const index_js_24 = require("./S1134/index.js");
const index_js_25 = require("./S1135/index.js");
const index_js_26 = require("./S1154/index.js");
const index_js_27 = require("./S1172/index.js");
const index_js_28 = require("./S1192/index.js");
const index_js_29 = require("./S1219/index.js");
const index_js_30 = require("./S1226/index.js");
const index_js_31 = require("./S1264/index.js");
const index_js_32 = require("./S1291/index.js");
const index_js_33 = require("./S1301/index.js");
const index_js_34 = require("./S1313/index.js");
const index_js_35 = require("./S1439/index.js");
const index_js_36 = require("./S1444/index.js");
const index_js_37 = require("./S1451/index.js");
const index_js_38 = require("./S1472/index.js");
const index_js_39 = require("./S1479/index.js");
const index_js_40 = require("./S1481/index.js");
const index_js_41 = require("./S1488/index.js");
const index_js_42 = require("./S1515/index.js");
const index_js_43 = require("./S1523/index.js");
const index_js_44 = require("./S1526/index.js");
const index_js_45 = require("./S1527/index.js");
const index_js_46 = require("./S1528/index.js");
const index_js_47 = require("./S1529/index.js");
const index_js_48 = require("./S1530/index.js");
const index_js_49 = require("./S1533/index.js");
const index_js_50 = require("./S1535/index.js");
const index_js_51 = require("./S1541/index.js");
const index_js_52 = require("./S1607/index.js");
const index_js_53 = require("./S1764/index.js");
const index_js_54 = require("./S1821/index.js");
const index_js_55 = require("./S1848/index.js");
const index_js_56 = require("./S1854/index.js");
const index_js_57 = require("./S1862/index.js");
const index_js_58 = require("./S1871/index.js");
const index_js_59 = require("./S1874/index.js");
const index_js_60 = require("./S1940/index.js");
const index_js_61 = require("./S1994/index.js");
const index_js_62 = require("./S2004/index.js");
const index_js_63 = require("./S2068/index.js");
const index_js_64 = require("./S2077/index.js");
const index_js_65 = require("./S2092/index.js");
const index_js_66 = require("./S2123/index.js");
const index_js_67 = require("./S2137/index.js");
const index_js_68 = require("./S2138/index.js");
const index_js_69 = require("./S2187/index.js");
const index_js_70 = require("./S2201/index.js");
const index_js_71 = require("./S2208/index.js");
const index_js_72 = require("./S2234/index.js");
const index_js_73 = require("./S2245/index.js");
const index_js_74 = require("./S2251/index.js");
const index_js_75 = require("./S2255/index.js");
const index_js_76 = require("./S2259/index.js");
const index_js_77 = require("./S2301/index.js");
const index_js_78 = require("./S2310/index.js");
const index_js_79 = require("./S2392/index.js");
const index_js_80 = require("./S2424/index.js");
const index_js_81 = require("./S2428/index.js");
const index_js_82 = require("./S2486/index.js");
const index_js_83 = require("./S2589/index.js");
const index_js_84 = require("./S2598/index.js");
const index_js_85 = require("./S2612/index.js");
const index_js_86 = require("./S2639/index.js");
const index_js_87 = require("./S2681/index.js");
const index_js_88 = require("./S2692/index.js");
const index_js_89 = require("./S2699/index.js");
const index_js_90 = require("./S2703/index.js");
const index_js_91 = require("./S2737/index.js");
const index_js_92 = require("./S2755/index.js");
const index_js_93 = require("./S2757/index.js");
const index_js_94 = require("./S2817/index.js");
const index_js_95 = require("./S2819/index.js");
const index_js_96 = require("./S2870/index.js");
const index_js_97 = require("./S2871/index.js");
const index_js_98 = require("./S2970/index.js");
const index_js_99 = require("./S2990/index.js");
const index_js_100 = require("./S2999/index.js");
const index_js_101 = require("./S3001/index.js");
const index_js_102 = require("./S3003/index.js");
const index_js_103 = require("./S3317/index.js");
const index_js_104 = require("./S3330/index.js");
const index_js_105 = require("./S3358/index.js");
const index_js_106 = require("./S3402/index.js");
const index_js_107 = require("./S3403/index.js");
const index_js_108 = require("./S3415/index.js");
const index_js_109 = require("./S3499/index.js");
const index_js_110 = require("./S3500/index.js");
const index_js_111 = require("./S3513/index.js");
const index_js_112 = require("./S3514/index.js");
const index_js_113 = require("./S3516/index.js");
const index_js_114 = require("./S3524/index.js");
const index_js_115 = require("./S3525/index.js");
const index_js_116 = require("./S3531/index.js");
const index_js_117 = require("./S3533/index.js");
const index_js_118 = require("./S3579/index.js");
const index_js_119 = require("./S3616/index.js");
const index_js_120 = require("./S3626/index.js");
const index_js_121 = require("./S3686/index.js");
const index_js_122 = require("./S3699/index.js");
const index_js_123 = require("./S3735/index.js");
const index_js_124 = require("./S3757/index.js");
const index_js_125 = require("./S3758/index.js");
const index_js_126 = require("./S3760/index.js");
const index_js_127 = require("./S3776/index.js");
const index_js_128 = require("./S3782/index.js");
const index_js_129 = require("./S3785/index.js");
const index_js_130 = require("./S3796/index.js");
const index_js_131 = require("./S3798/index.js");
const index_js_132 = require("./S3800/index.js");
const index_js_133 = require("./S3801/index.js");
const index_js_134 = require("./S3827/index.js");
const index_js_135 = require("./S3923/index.js");
const index_js_136 = require("./S3972/index.js");
const index_js_137 = require("./S3973/index.js");
const index_js_138 = require("./S3981/index.js");
const index_js_139 = require("./S3984/index.js");
const index_js_140 = require("./S4030/index.js");
const index_js_141 = require("./S4036/index.js");
const index_js_142 = require("./S4043/index.js");
const index_js_143 = require("./S4139/index.js");
const index_js_144 = require("./S4143/index.js");
const index_js_145 = require("./S4144/index.js");
const index_js_146 = require("./S4158/index.js");
const index_js_147 = require("./S4165/index.js");
const index_js_148 = require("./S4322/index.js");
const index_js_149 = require("./S4323/index.js");
const index_js_150 = require("./S4324/index.js");
const index_js_151 = require("./S4328/index.js");
const index_js_152 = require("./S4335/index.js");
const index_js_153 = require("./S4423/index.js");
const index_js_154 = require("./S4426/index.js");
const index_js_155 = require("./S4502/index.js");
const index_js_156 = require("./S4507/index.js");
const index_js_157 = require("./S4524/index.js");
const index_js_158 = require("./S4619/index.js");
const index_js_159 = require("./S4621/index.js");
const index_js_160 = require("./S4622/index.js");
const index_js_161 = require("./S4623/index.js");
const index_js_162 = require("./S4624/index.js");
const index_js_163 = require("./S4634/index.js");
const index_js_164 = require("./S4721/index.js");
const index_js_165 = require("./S4782/index.js");
const index_js_166 = require("./S4784/index.js");
const index_js_167 = require("./S4787/index.js");
const index_js_168 = require("./S4790/index.js");
const index_js_169 = require("./S4798/index.js");
const index_js_170 = require("./S4817/index.js");
const index_js_171 = require("./S4818/index.js");
const index_js_172 = require("./S4822/index.js");
const index_js_173 = require("./S4823/index.js");
const index_js_174 = require("./S4829/index.js");
const index_js_175 = require("./S4830/index.js");
const index_js_176 = require("./S5042/index.js");
const index_js_177 = require("./S5122/index.js");
const index_js_178 = require("./S5148/index.js");
const index_js_179 = require("./S5247/index.js");
const index_js_180 = require("./S5256/index.js");
const index_js_181 = require("./S5257/index.js");
const index_js_182 = require("./S5260/index.js");
const index_js_183 = require("./S5264/index.js");
const index_js_184 = require("./S5332/index.js");
const index_js_185 = require("./S5443/index.js");
const index_js_186 = require("./S5527/index.js");
const index_js_187 = require("./S5542/index.js");
const index_js_188 = require("./S5547/index.js");
const index_js_189 = require("./S5604/index.js");
const index_js_190 = require("./S5659/index.js");
const index_js_191 = require("./S5689/index.js");
const index_js_192 = require("./S5691/index.js");
const index_js_193 = require("./S5693/index.js");
const index_js_194 = require("./S5725/index.js");
const index_js_195 = require("./S5728/index.js");
const index_js_196 = require("./S5730/index.js");
const index_js_197 = require("./S5732/index.js");
const index_js_198 = require("./S5734/index.js");
const index_js_199 = require("./S5736/index.js");
const index_js_200 = require("./S5739/index.js");
const index_js_201 = require("./S5742/index.js");
const index_js_202 = require("./S5743/index.js");
const index_js_203 = require("./S5757/index.js");
const index_js_204 = require("./S5759/index.js");
const index_js_205 = require("./S5842/index.js");
const index_js_206 = require("./S5843/index.js");
const index_js_207 = require("./S5850/index.js");
const index_js_208 = require("./S5852/index.js");
const index_js_209 = require("./S5856/index.js");
const index_js_210 = require("./S5860/index.js");
const index_js_211 = require("./S5863/index.js");
const index_js_212 = require("./S5867/index.js");
const index_js_213 = require("./S5868/index.js");
const index_js_214 = require("./S5869/index.js");
const index_js_215 = require("./S5876/index.js");
const index_js_216 = require("./S5958/index.js");
const index_js_217 = require("./S5973/index.js");
const index_js_218 = require("./S6019/index.js");
const index_js_219 = require("./S6035/index.js");
const index_js_220 = require("./S6079/index.js");
const index_js_221 = require("./S6080/index.js");
const index_js_222 = require("./S6092/index.js");
const index_js_223 = require("./S6245/index.js");
const index_js_224 = require("./S6249/index.js");
const index_js_225 = require("./S6252/index.js");
const index_js_226 = require("./S6265/index.js");
const index_js_227 = require("./S6268/index.js");
const index_js_228 = require("./S6270/index.js");
const index_js_229 = require("./S6275/index.js");
const index_js_230 = require("./S6281/index.js");
const index_js_231 = require("./S6299/index.js");
const index_js_232 = require("./S6302/index.js");
const index_js_233 = require("./S6303/index.js");
const index_js_234 = require("./S6304/index.js");
const index_js_235 = require("./S6308/index.js");
const index_js_236 = require("./S6317/index.js");
const index_js_237 = require("./S6319/index.js");
const index_js_238 = require("./S6321/index.js");
const index_js_239 = require("./S6323/index.js");
const index_js_240 = require("./S6324/index.js");
const index_js_241 = require("./S6326/index.js");
const index_js_242 = require("./S6327/index.js");
const index_js_243 = require("./S6328/index.js");
const index_js_244 = require("./S6329/index.js");
const index_js_245 = require("./S6330/index.js");
const index_js_246 = require("./S6331/index.js");
const index_js_247 = require("./S6332/index.js");
const index_js_248 = require("./S6333/index.js");
const index_js_249 = require("./S6351/index.js");
const index_js_250 = require("./S6353/index.js");
const index_js_251 = require("./S6397/index.js");
const index_js_252 = require("./S6418/index.js");
const index_js_253 = require("./S6426/index.js");
const index_js_254 = require("./S6437/index.js");
const index_js_255 = require("./S6439/index.js");
const index_js_256 = require("./S6442/index.js");
const index_js_257 = require("./S6443/index.js");
const index_js_258 = require("./S6486/index.js");
const index_js_259 = require("./S6564/index.js");
const index_js_260 = require("./S6594/index.js");
const index_js_261 = require("./S6627/index.js");
const index_js_262 = require("./S6759/index.js");
const index_js_263 = require("./S6958/index.js");
const index_js_264 = require("./S6959/index.js");
const index_js_265 = require("./S7059/index.js");
const index_js_266 = require("./S7639/index.js");
const index_js_267 = require("./S7790/index.js");
const index_js_268 = require("./S8441/index.js");
const index_js_269 = require("./S8479/index.js");
exports.rules = {
    'function-name': index_js_1.rule,
    'class-name': index_js_2.rule,
    'max-lines': index_js_3.rule,
    'no-tab': index_js_4.rule,
    'variable-name': index_js_5.rule,
    'comment-regex': index_js_6.rule,
    'no-commented-code': index_js_7.rule,
    'elseif-without-else': index_js_8.rule,
    'no-fallthrough': index_js_9.rule,
    'nested-control-flow': index_js_10.rule,
    'too-many-break-or-continue-in-loop': index_js_11.rule,
    'max-lines-per-function': index_js_12.rule,
    'no-nested-incdec': index_js_13.rule,
    'no-equals-in-for-termination': index_js_14.rule,
    'no-extra-arguments': index_js_15.rule,
    'no-collapsible-if': index_js_16.rule,
    'expression-complexity': index_js_17.rule,
    'no-redundant-parentheses': index_js_18.rule,
    'no-labels': index_js_19.rule,
    'no-nested-assignment': index_js_20.rule,
    'no-redundant-boolean': index_js_21.rule,
    'prefer-single-boolean-return': index_js_22.rule,
    'unused-import': index_js_23.rule,
    'fixme-tag': index_js_24.rule,
    'todo-tag': index_js_25.rule,
    'useless-string-operation': index_js_26.rule,
    'no-unused-function-argument': index_js_27.rule,
    'no-duplicate-string': index_js_28.rule,
    'no-case-label-in-switch': index_js_29.rule,
    'no-parameter-reassignment': index_js_30.rule,
    'prefer-while': index_js_31.rule,
    'no-sonar-comments': index_js_32.rule,
    'no-small-switch': index_js_33.rule,
    'no-hardcoded-ip': index_js_34.rule,
    'label-position': index_js_35.rule,
    'public-static-readonly': index_js_36.rule,
    'file-header': index_js_37.rule,
    'call-argument-line': index_js_38.rule,
    'max-switch-cases': index_js_39.rule,
    'no-unused-vars': index_js_40.rule,
    'prefer-immediate-return': index_js_41.rule,
    'function-inside-loop': index_js_42.rule,
    'code-eval': index_js_43.rule,
    'no-variable-usage-before-declaration': index_js_44.rule,
    'future-reserved-words': index_js_45.rule,
    'array-constructor': index_js_46.rule,
    'bitwise-operators': index_js_47.rule,
    'no-function-declaration-in-block': index_js_48.rule,
    'no-primitive-wrappers': index_js_49.rule,
    'for-in': index_js_50.rule,
    'cyclomatic-complexity': index_js_51.rule,
    'no-skipped-tests': index_js_52.rule,
    'no-identical-expressions': index_js_53.rule,
    'no-nested-switch': index_js_54.rule,
    'constructor-for-side-effects': index_js_55.rule,
    'no-dead-store': index_js_56.rule,
    'no-identical-conditions': index_js_57.rule,
    'no-duplicated-branches': index_js_58.rule,
    deprecation: index_js_59.rule,
    'no-inverted-boolean-check': index_js_60.rule,
    'misplaced-loop-counter': index_js_61.rule,
    'no-nested-functions': index_js_62.rule,
    'no-hardcoded-passwords': index_js_63.rule,
    'sql-queries': index_js_64.rule,
    'insecure-cookie': index_js_65.rule,
    'no-useless-increment': index_js_66.rule,
    'no-globals-shadowing': index_js_67.rule,
    'no-undefined-assignment': index_js_68.rule,
    'no-empty-test-file': index_js_69.rule,
    'no-ignored-return': index_js_70.rule,
    'no-wildcard-import': index_js_71.rule,
    'arguments-order': index_js_72.rule,
    'pseudo-random': index_js_73.rule,
    'for-loop-increment-sign': index_js_74.rule,
    cookies: index_js_75.rule,
    'null-dereference': index_js_76.rule,
    'no-selector-parameter': index_js_77.rule,
    'updated-loop-counter': index_js_78.rule,
    'block-scoped-var': index_js_79.rule,
    'no-built-in-override': index_js_80.rule,
    'prefer-object-literal': index_js_81.rule,
    'no-ignored-exceptions': index_js_82.rule,
    'no-gratuitous-expressions': index_js_83.rule,
    'file-uploads': index_js_84.rule,
    'file-permissions': index_js_85.rule,
    'no-empty-character-class': index_js_86.rule,
    'no-unenclosed-multiline-block': index_js_87.rule,
    'index-of-compare-to-positive-number': index_js_88.rule,
    'assertions-in-tests': index_js_89.rule,
    'no-implicit-global': index_js_90.rule,
    'no-useless-catch': index_js_91.rule,
    'xml-parser-xxe': index_js_92.rule,
    'non-existent-operator': index_js_93.rule,
    'web-sql-database': index_js_94.rule,
    'post-message': index_js_95.rule,
    'no-array-delete': index_js_96.rule,
    'no-alphabetical-sort': index_js_97.rule,
    'no-incomplete-assertions': index_js_98.rule,
    'no-global-this': index_js_99.rule,
    'new-operator-misuse': index_js_100.rule,
    'no-delete-var': index_js_101.rule,
    'strings-comparison': index_js_102.rule,
    'file-name-differ-from-class': index_js_103.rule,
    'cookie-no-httponly': index_js_104.rule,
    'no-nested-conditional': index_js_105.rule,
    'no-incorrect-string-concat': index_js_106.rule,
    'different-types-comparison': index_js_107.rule,
    'inverted-assertion-arguments': index_js_108.rule,
    'shorthand-property-grouping': index_js_109.rule,
    'updated-const-var': index_js_110.rule,
    'arguments-usage': index_js_111.rule,
    'destructuring-assignment-syntax': index_js_112.rule,
    'no-invariant-returns': index_js_113.rule,
    'arrow-function-convention': index_js_114.rule,
    'class-prototype': index_js_115.rule,
    'generator-without-yield': index_js_116.rule,
    'no-require-or-define': index_js_117.rule,
    'no-associative-arrays': index_js_118.rule,
    'comma-or-logical-or-case': index_js_119.rule,
    'no-redundant-jump': index_js_120.rule,
    'inconsistent-function-call': index_js_121.rule,
    'no-use-of-empty-return-value': index_js_122.rule,
    'void-use': index_js_123.rule,
    'operation-returning-nan': index_js_124.rule,
    'values-not-convertible-to-numbers': index_js_125.rule,
    'non-number-in-arithmetic-expression': index_js_126.rule,
    'cognitive-complexity': index_js_127.rule,
    'argument-type': index_js_128.rule,
    'in-operator-type-error': index_js_129.rule,
    'array-callback-without-return': index_js_130.rule,
    'declarations-in-global-scope': index_js_131.rule,
    'function-return-type': index_js_132.rule,
    'no-inconsistent-returns': index_js_133.rule,
    'no-reference-error': index_js_134.rule,
    'no-all-duplicated-branches': index_js_135.rule,
    'no-same-line-conditional': index_js_136.rule,
    'conditional-indentation': index_js_137.rule,
    'no-collection-size-mischeck': index_js_138.rule,
    'no-unthrown-error': index_js_139.rule,
    'no-unused-collection': index_js_140.rule,
    'no-os-command-from-path': index_js_141.rule,
    'no-misleading-array-reverse': index_js_142.rule,
    'no-for-in-iterable': index_js_143.rule,
    'no-element-overwrite': index_js_144.rule,
    'no-identical-functions': index_js_145.rule,
    'no-empty-collection': index_js_146.rule,
    'no-redundant-assignments': index_js_147.rule,
    'prefer-type-guard': index_js_148.rule,
    'use-type-alias': index_js_149.rule,
    'no-return-type-any': index_js_150.rule,
    'no-implicit-dependencies': index_js_151.rule,
    'no-useless-intersection': index_js_152.rule,
    'weak-ssl': index_js_153.rule,
    'no-weak-keys': index_js_154.rule,
    csrf: index_js_155.rule,
    'production-debug': index_js_156.rule,
    'prefer-default-last': index_js_157.rule,
    'no-in-misuse': index_js_158.rule,
    'no-duplicate-in-composite': index_js_159.rule,
    'max-union-size': index_js_160.rule,
    'no-undefined-argument': index_js_161.rule,
    'no-nested-template-literals': index_js_162.rule,
    'prefer-promise-shorthand': index_js_163.rule,
    'os-command': index_js_164.rule,
    'no-redundant-optional': index_js_165.rule,
    'regular-expr': index_js_166.rule,
    encryption: index_js_167.rule,
    hashing: index_js_168.rule,
    'bool-param-default': index_js_169.rule,
    xpath: index_js_170.rule,
    sockets: index_js_171.rule,
    'no-try-promise': index_js_172.rule,
    'process-argv': index_js_173.rule,
    'standard-input': index_js_174.rule,
    'unverified-certificate': index_js_175.rule,
    'no-unsafe-unzip': index_js_176.rule,
    cors: index_js_177.rule,
    'link-with-target-blank': index_js_178.rule,
    'disabled-auto-escaping': index_js_179.rule,
    'table-header': index_js_180.rule,
    'no-table-as-layout': index_js_181.rule,
    'table-header-reference': index_js_182.rule,
    'object-alt-content': index_js_183.rule,
    'no-clear-text-protocols': index_js_184.rule,
    'publicly-writable-directories': index_js_185.rule,
    'unverified-hostname': index_js_186.rule,
    'encryption-secure-mode': index_js_187.rule,
    'no-weak-cipher': index_js_188.rule,
    'no-intrusive-permissions': index_js_189.rule,
    'insecure-jwt-token': index_js_190.rule,
    'x-powered-by': index_js_191.rule,
    'hidden-files': index_js_192.rule,
    'content-length': index_js_193.rule,
    'disabled-resource-integrity': index_js_194.rule,
    'content-security-policy': index_js_195.rule,
    'no-mixed-content': index_js_196.rule,
    'frame-ancestors': index_js_197.rule,
    'no-mime-sniff': index_js_198.rule,
    'no-referrer-policy': index_js_199.rule,
    'strict-transport-security': index_js_200.rule,
    'certificate-transparency': index_js_201.rule,
    'dns-prefetching': index_js_202.rule,
    'confidential-information-logging': index_js_203.rule,
    'no-ip-forward': index_js_204.rule,
    'empty-string-repetition': index_js_205.rule,
    'regex-complexity': index_js_206.rule,
    'anchor-precedence': index_js_207.rule,
    'slow-regex': index_js_208.rule,
    'no-invalid-regexp': index_js_209.rule,
    'unused-named-groups': index_js_210.rule,
    'no-same-argument-assert': index_js_211.rule,
    'unicode-aware-regex': index_js_212.rule,
    'no-misleading-character-class': index_js_213.rule,
    'duplicates-in-character-class': index_js_214.rule,
    'session-regeneration': index_js_215.rule,
    'test-check-exception': index_js_216.rule,
    'stable-tests': index_js_217.rule,
    'no-empty-after-reluctant': index_js_218.rule,
    'single-character-alternation': index_js_219.rule,
    'no-code-after-done': index_js_220.rule,
    'disabled-timeout': index_js_221.rule,
    'chai-determinate-assertion': index_js_222.rule,
    'aws-s3-bucket-server-encryption': index_js_223.rule,
    'aws-s3-bucket-insecure-http': index_js_224.rule,
    'aws-s3-bucket-versioning': index_js_225.rule,
    'aws-s3-bucket-granted-access': index_js_226.rule,
    'no-angular-bypass-sanitization': index_js_227.rule,
    'aws-iam-public-access': index_js_228.rule,
    'aws-ec2-unencrypted-ebs-volume': index_js_229.rule,
    'aws-s3-bucket-public-access': index_js_230.rule,
    'no-vue-bypass-sanitization': index_js_231.rule,
    'aws-iam-all-privileges': index_js_232.rule,
    'aws-rds-unencrypted-databases': index_js_233.rule,
    'aws-iam-all-resources-accessible': index_js_234.rule,
    'aws-opensearchservice-domain': index_js_235.rule,
    'aws-iam-privilege-escalation': index_js_236.rule,
    'aws-sagemaker-unencrypted-notebook': index_js_237.rule,
    'aws-restricted-ip-admin-access': index_js_238.rule,
    'no-empty-alternatives': index_js_239.rule,
    'no-control-regex': index_js_240.rule,
    'no-regex-spaces': index_js_241.rule,
    'aws-sns-unencrypted-topics': index_js_242.rule,
    'existing-groups': index_js_243.rule,
    'aws-ec2-rds-dms-public': index_js_244.rule,
    'aws-sqs-unencrypted-queue': index_js_245.rule,
    'no-empty-group': index_js_246.rule,
    'aws-efs-unencrypted': index_js_247.rule,
    'aws-apigateway-public-api': index_js_248.rule,
    'stateful-regex': index_js_249.rule,
    'concise-regex': index_js_250.rule,
    'single-char-in-character-classes': index_js_251.rule,
    'no-hardcoded-secrets': index_js_252.rule,
    'no-exclusive-tests': index_js_253.rule,
    'hardcoded-secret-signatures': index_js_254.rule,
    'jsx-no-leaked-render': index_js_255.rule,
    'no-hook-setter-in-body': index_js_256.rule,
    'no-useless-react-setstate': index_js_257.rule,
    'no-uniq-key': index_js_258.rule,
    'redundant-type-aliases': index_js_259.rule,
    'prefer-regexp-exec': index_js_260.rule,
    'no-internal-api-use': index_js_261.rule,
    'prefer-read-only-props': index_js_262.rule,
    'no-literal-call': index_js_263.rule,
    'reduce-initial-value': index_js_264.rule,
    'no-async-constructor': index_js_265.rule,
    'review-blockchain-mnemonic': index_js_266.rule,
    'dynamically-constructed-templates': index_js_267.rule,
    'no-session-cookies-on-static-assets': index_js_268.rule,
    'dompurify-unsafe-config': index_js_269.rule,
};
